Ask new input when input is invalid

carakas opened the pull request:
I altered the FormHelper a bit so it will keep asking input for the invalid form fields until all the data is valid. Then it will return it.

If there is already data for the collection the form will loop over them so you can edit them.
If deleting is allowed you can also choose not to add the item.

To make it easier to follow the current count is also displayed each time an item is edited or added

As a side effect all forms now take predefined data

screenshot of this code in action
<img width="428" alt="screen shot 2017-01-30 at 22 25 22" src="https://cloud.githubusercontent.com/assets/3634654/22442272/0e0db65a-e73b-11e6-8db7-c33bdc8d9c8a.png">

- [x] Ask to fill in the input again when it is invalid
- [x] Allow initial data to be passed to the form
- [x] Handle pre-existing data in collections instead of overwriting them
- [x] Write tests

carakas committed:
Ask new input when input is invalid

matthiasnoback commented:
Hi! Thanks for submitting this PR. I absolutely love the way this is heading! However, reading the code, I wonder: does this work for nested forms too, e.g. collection-type forms?

matthiasnoback commented:
Also, could you add an example for this behavior to the Behat features?

carakas commented:
Good, question. I'm currently implementing a module generator with this. I'll check tonight how it handles collections.

For the Behat, I'll try :) never worked with it before

matthiasnoback commented:
Awesome, great work!

carakas commented:
It throws the error if the collection is not valid but then doesn't ask it again.
I'll see what I can do to make it also work with collections

Edit: it does ask it again if I set `error_bubbling` to `false` on the `CollectionType` but the data of the `entry_type` is still lost at the moment

carakas commented:
@matthiasnoback I did some testing, is it possible that the current implementation for collections can't handle pre set data?

```
ModuleGenerator\Domain\ValueObject\ValueObjectDataTransferObject {#169
  +namespace: ModuleGenerator\PhpGenerator\DataTransferObject\NamespaceDataTransferObject {#303
    +name: "\Console"
  }
  +className: "Status"
  +constants: array:2 [
    0 => ModuleGenerator\PhpGenerator\DataTransferObject\ConstantDataTransferObject {#508
      +value: "the builder"
      +name: "Bob"
      +visibility: null
      +comment: null
    }
    1 => ModuleGenerator\PhpGenerator\DataTransferObject\ConstantDataTransferObject {#534
      +value: null
      +name: "bob"
      +visibility: null
      +comment: null
    }
  ]
}
Invalid data provided: constants:
    1:
        value:
            ERROR: This value should not be blank.
array:2 [
  0 => "namespace"
  1 => "className"
]
Constants
Add another entry to this collection? [n]: n
ModuleGenerator\Domain\ValueObject\ValueObjectDataTransferObject {#169
  +namespace: ModuleGenerator\PhpGenerator\DataTransferObject\NamespaceDataTransferObject {#303
    +name: "\Console"
  }
  +className: "Status"
  +constants: []
}
Invalid data provided: constants:
    ERROR: You need to add at least 1 constant
array:2 [
  0 => "namespace"
  1 => "className"
]
```

Do you have some suggestions on how to handle it if it's indeed the case?

matthiasnoback commented:
Hmm, not sure if that works actually :) I never tried to render the form again.

carakas commented:
Is it something that should be added? If so, do I need to create a separate PR for that?

matthiasnoback commented:
Of course, it would be great to add this feature to this library! Doesn't have to be a separate PR I think.

carakas commented:
ok, I'll work on it when I have time

matthiasnoback commented:
Excellent, we're not in a hurry I think ;)

carakas committed:
Add missing dev requirement

The symfony/security was missing

carakas committed:
Allow setting the initial data when interacting with the form

carakas commented:
I'm currently adding the implementation details, the test will follow later, I've never worked with behat before so I have some things to figure out but have no time for that atm

matthiasnoback commented:
Nice :) 
I guess you shouldn't need to know a lot about Behat. All the tests are like: "using *this* form, when I enter *this* data, I'll expect this to be printed on the screen".

carakas committed:
Allow the iterating over existing data

As a side effect of this data previously submitted in a collection that wasn't completely valid will no longer be lost

carakas committed:
The allow add is only needed if the collections isn't prefilled

carakas committed:
Add test for resubmitting the invalid data

carakas committed:
Make it possible to check that something is not in the output

carakas committed:
Add test for prefilled data in collections

carakas committed:
Print counter in collection and update the readme

carakas commented:
@matthiasnoback I've added the tests and completed the implementation

carakas committed:
Fix data of the last edited collection entry being used as the default for the new entries

matthiasnoback commented:
👏 👏 👏 👏 👏 👏 👏 👏 
I think you've put a lot of work into this, thanks!
Will merge it in a minute.

tarlepp commented:
Hmm, just updated to new version and it seems that 
```
$builder
    ->add(
        'username',
        Type\TextType::class,
        [
            'label'     => 'Username',
            'required'  => true,
        ]
    )
```

does not trigger asking again that input and it accepts just empty value.

carakas commented:
@tarlepp is this inside a collection or just on the base form?

tarlepp commented:
@carakas that one https://github.com/tarlepp/symfony-backend/blob/master/src/App/Form/Console/User.php

carakas commented:
the required attribute only triggers the html5 validation afaik, and that does nothing in the console
you need to add a contraint to it like this
```php
$builder->add(
            'name',
            TextType::class,
            [
                'label' => $options['name_label'],
                'required' => true,
                'constraints' => [
                    new NotBlank(),
                    new Regex(
                        [
                            'pattern' => '/^[a-zA-Z_\\x7f-\\xff][a-zA-Z0-9_\\x7f-\\xff]*$/',
                            'message' => 'Invalid class name',
                        ]
                    ),
                ],
            ]
        );
```

on collections you need to add
```php
'constraints' => [
    new Valid()
]
```

tarlepp commented:
Hmm... Tried

```
use Symfony\Component\Validator\Constraints as Assert;

$builder
            ->add(
                'username',
                Type\TextType::class,
                [
                    'label'         => 'Username',
                    'required'      => true,
                    'constraints'   => [
                        new Assert\NotBlank(),
                        new Assert\NotNull(),
                        new Assert\Length(['min' => 2, 'max' => 255]),
                    ]
                ]
            )
```

But still same behaviour.

carakas commented:
what happens if you set `'error_bubbling' => false` on your form as default?


```php
public function configureOptions(OptionsResolver $resolver)
    {
        $resolver->setDefaults(
            [
                'error_bubbling' => false,
            ]
        );
    }
```

carakas commented:
cause it is working here :)

carakas commented:
@tarlepp have you checked https://github.com/matthiasnoback/symfony-console-form/pull/21/files#diff-d263646a686fcef5c7a9b6d486c03f82 ? because this one works in the tests

tarlepp commented:
tried that error_bubbling, no results.

carakas commented:
@tarlepp just tested your code and it does seem to work on createGroup

tarlepp commented:
I'm trying to get that work, https://gist.github.com/tarlepp/cab838bc9ad319c647c17ed5f493ccd8 this doesn't work - could you share your implementation ?

carakas commented:
@tarlepp I think I found out what the issue is
your user DTO has methods like getUsername, I think those collide with the assertions that you have added to the public property username

are you sure you are on 2.2.0?
<img width="355" alt="screen shot 2017-02-02 at 18 53 51" src="https://cloud.githubusercontent.com/assets/3634654/22561538/f54bc78e-e978-11e6-8375-848e39ef7ddd.png">
I didn't change anything to the usergroup implementation, it just worked

tarlepp commented:
fuuuuck - I'm just retarded... I though it will do that validation after you hit enter on that field...

@carakas really sorry about this mess :D

carakas commented:
:) it works the same way as a normal implementation, on submit :)
